### PR TITLE
[WIP] remove waitpub, export publish

### PR DESCRIPTION
**Pull request by schomatis**
Fixes https://github.com/ipfs/go-mfs/issues/38.

### CODE REVIEW

**Comment by schomatis** [comment]
@Stebalien This is a rough sketch of my proposal in #38, could you take a look at it please and tell me what you think please?

This patch doesn't guarantee that different `PublishNow` will actually happen in the order they were called (but this isn't guaranteed in the current code anyways, since the order of `Update` calls isn't enforced).

File: repub.go [hunk 1]
```diff
@@ -18,11 +18,9 @@ type Republisher struct {
 	TimeoutShort        time.Duration
 	valueHasBeenUpdated chan struct{}
 	pubfunc             PubFunc
-	immediatePublish    chan chan struct{}
 
 	valueLock          sync.Mutex
-	valueToPublish     cid.Cid
-	lastValuePublished cid.Cid
+	valueToPublish     *cid.Cid
```
[the remark this answers is not in the document]
Comment (nitishm) [reply repub.go]: Ah I think I see how this handled in `PublishNow()` with the `extractedValue == nil` check.

[the remark this answers is not in the document]
Comment (Stebalien) [reply repub.go]: This should just be a `cid.Cid` and should be set to `cid.Undef` when "nil". (probably should have been `cid.Nil` but I didn't win that argument).

File: repub.go [hunk 2]
```diff
@@ -139,17 +113,20 @@ func (rp *Republisher) Run() {
...
-	topub := rp.valueToPublish
+	extractedValue := rp.valueToPublish
+	rp.valueToPublish = nil
 	rp.valueLock.Unlock()
 
-	err := rp.pubfunc(ctx, topub)
-	if err != nil {
-		return err
+	if extractedValue == nil {
+		return nil
```
Comment (Stebalien) [inline repub.go]: A concurrent call won't actually wait. We may need a RwMutex here.

File: repub.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (schomatis) [reply repub.go]: Yes, I thought of that at first but we *do* `cid.Undef` to issue publish orders, at least in some tests,

https://github.com/ipfs/go-mfs/blob/4fb6dc421ff8f41bab540bd91654e0b33ee41ad6/repub_test.go#L43-L45

so I think it violates the semantics I would expect of a `nil` value.

I agree that we should fix that and provide a `cid.Nil` but in the meanwhile I don't see the harm in implementing the `nil` with a pointer for an *internal* variable.

**Comment by schomatis** [comment]
> However, I'd expect `PublishNow` to only return _after_ the latest value has been published (even if there's a concurrent publish).

Yes, that seems fair but `WaitPub` didn't do that, and my main objective here is to simplify the code, not redefine behavior, maybe `PublishNow` is a misleading name, would `PublishCurrentValue` be better?

**Comment by Stebalien** [comment]
> Yes, that seems fair but WaitPub didn't do that, and my main objective here is to simplify the code, not redefine behavior, maybe PublishNow is a misleading name, would PublishCurrentValue be better?

If one changes a file and then calls `WaitPub`, `WaitPub` is guaranteed to not return until that change has been published. Of course, given multiple calls to `WaitPub`, one (or more) of these calls may block waiting for yet another change (that's the issue you're fixing).

Here, given two commands call `PublishNow` at the same time, one command will return early (before the publish happens). That means `PublishNow`, as implemented, isn't useful as a replacement for `WaitPub`.

File: repub.go [hunk 3]
```diff
@@ -139,17 +113,20 @@ func (rp *Republisher) Run() {
 
 // Wrapper function around the user-defined `pubfunc`. It publishes
 // the (last) `valueToPublish` set and registers it in `lastValuePublished`.
-func (rp *Republisher) publish(ctx context.Context) error {
+// TODO: Allow passing a value to `PublishNow` which supersedes the
+// internal `valueToPublish`.
```
Comment (Stebalien) [inline repub.go]: I'm not sure if we want to allow this. Users shouldn't swap out the MFS root using the republisher.

Comment (schomatis) [reply repub.go]: I'm not sure I understand the comment, what do you mean by swap out?

The `TODO` (that I don't think I worded correctly) was aiming at adding an optional argument that would replace the `Update(newCid); PublishNow();` call pair with just a `PublishNow(newCid)` call.

Comment (Stebalien) [reply repub.go]: Got it. Yeah, that makes sense.

(context: I keep thinking we're exposing the republisher to the user)

File: repub.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Stebalien) [reply repub.go]: The pointer is fine but shouldn't be necessary. Really, we should fix that test, passing the "Undef" CID is should be equivalent to a "nil" CID (and passing a nil CID to `rp.Update` doesn't make sense).

**Comment by Stebalien** [comment]
> does that sound fair to you?

Are you referring to your latest change? That's going to overwrite potentially unpublished values. Also, we really shouldn't be allowing the user to invoke `pubfunc` multiple times in parallel (both for thread safety and to prevent logical races where we publish values out of order). In the past, this was protected by the loop.

**Comment by Stebalien** [comment]
> This is what I was wondering in #38 (comment), but from your following comments I assumed this wasn't a problem, if I misunderstood that (sorry!) and we can't risk calling it in parallel then we should just close this PR, which has no way of working then.

It's probably thread-safe (although we shouldn't assume that) however, that doesn't matter. If the user calls `Update(x)` and then `Update(y)` in a single thread, `y` should win. With the new code, `x` could win (and could even *prevent* `y` from being published) given a concurrent `PublishNow()` call.

> This is actually what motivated this PR in the first place. 

Given two simultaneous calls to `Update`, one will always trump the other. However, given two sequential calls to update with interleaved calls to `WaitPub` and/or `PublishNow`, the last published value should always win. That's the real issue here (the end-user can't currently call `Update` *anyways*).

> Anyway, if my appreciation in either of those points is wrong let's close this PR, my only objective at the moment, since I don't have much time left for proper fixes and redesigns, is to simplify the code for the next one to come along (hopefully not you :) to get a more clear perspective of what the code does (and doesn't do) and how could that be improved upon. Any suggestions towards that end you could propose I'll try to implement during next week.

Fixing #38 shouldn't require a redesign.

**Comment by schomatis** [comment]
> (If so, could you expand on why?)

Ok, I think I get it, the loop in `Run` doesn't play nice with independent `PublishNow()` calls from the user.

**Comment by schomatis** [comment]
Holding the lock throughout `pubfunc` seems like too much, closing then.

**Pull request closed without merging**